docs/proposals/nodeselector.md

```diff
@@ -0,0 +1,174 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/nodeselector.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NodeSelector, NodeConstraints, and NodePreferences
+
+## Introduction
+
+This document proposes a new label selector representation, called `NodeSelector`,
+that is similar in many ways to `PodSelector`, but is a bit more flexible and is
+intended to be used for selecting nodes rather than pods.
+
+In addition, we propose to replace the `map[string]string` in `PodSpec` that the scheduler
+currently uses as part of restricting the set of nodes onto which a pod is
+eligible to schedule, with a field called `NodeConstraints` that wil be represented
+by a `NodeSelector`. (In practice, as discussed later,
+we will *add* the `NodeConstraints` field rather than replacing `map[string]string`,
+due to backward compatibility requirements.) We further propose an additional
+new field for `PodSpec`, called `NodePreferences`, that specifies which nodes are
+preferred for scheduling among those that meet the hard requirements described
+by `NodeConstraints`.
+
+`NodeConstraints` and `NodePreferences` allow a pod to request various properties
+that are inherent to nodes, for example "run this pod on a node with an Intel CPU" or, in a
+multi-zone cluster, "run this pod on a node in zone Z."
+([This issue](https://github.com/kubernetes/kubernetes/issues/9044) describes
+some of the properties that a node might publish as labels, which `NodeConstraints`
+and `NodePreferences` can match against.)
+They do *not* allow a pod to request to schedule
+(or not schedule) on a node based on what other pods are running on the node. That
+feature is called "inter-pod topological affinity/anti-afinity" and is described
+[here](https://github.com/kubernetes/kubernetes/pull/18265).
+
+## API
+
+### NodeSelector
+
+```go
+// A node selector represents the union of the results of one or more label queries
+// over a set of nodes; that is, it represents the OR of the selectors represented
+// by the nodeSelectorTerms.
+// TODO: what are the semantics of empty and null?
+type NodeSelector struct {
+	// nodeSelectorTerms is a list of node selector terms. The terms are ORed.
```

Comment:
> Why the duplication between JobSelector and this new NodeSelector?

The answer to this is exactly what @sdminonne said.

> I still don't understand why NodeSelector is specialized.

We have two choices:
(1) represent node selection using LabelSelector (used to be called PodSelector)
(2) represent node selection using something new

I feel strongly that for node selection, we need Gt and Lt, and somewhat less strongly that we need top-level OR. This means that for (1) we'd need need to add Gt/Lt and top-level OR to LabelSelector. @bgrant0607 was very opposed to this -- I believe his reason was that you want to build a standard set of forward and reverse indexes for LabelSelector because lookup speed is very important for selection operations related to pods, and having more complex expressions (like Gt/Lt) makes such implementation more complicated. The only option that satisfies his and my requirements is (2), hence NodeSelector as a separate type.

(Example use case for Gt is "OS version is greater than ...")